lib/Server.js

```diff
@@ -501,24 +428,20 @@ class Server {
       runnableFeatures.push('proxy', 'middleware');
     }
 
-    if (this.options.contentBase !== false) {
-      runnableFeatures.push('contentBaseFiles');
+    if (this.options.static) {
+      runnableFeatures.push('static');
     }
 
     if (this.options.historyApiFallback) {
       runnableFeatures.push('historyApiFallback', 'middleware');
 
-      if (this.options.contentBase !== false) {
-        runnableFeatures.push('contentBaseFiles');
+      if (this.options.static) {
+        runnableFeatures.push('static');
       }
     }
 
-    if (this.options.contentBase && this.options.serveIndex) {
-      runnableFeatures.push('contentBaseIndex');
-    }
-
-    if (this.options.watchContentBase) {
-      runnableFeatures.push('watchContentBase');
+    if (this.options.static) {
+      runnableFeatures.push('serveIndex', 'watchStatic');
```

Comment:
I think we can union `static`/`serveIndex`/`watchStatic` in one method and middleware, using one without the other doesn't make sense at all

Reply:
@evilebottnawi I tried that but it causes breaking changes when `historyApiFallback` is enabled. It seems that the order of middleware things that are enabled surrounding historyApiFallback is important

https://github.com/webpack/webpack-dev-server/blob/master/lib/Server.js#L554-L575

Reply:
Why do nut put `static` feature after `historyApiFallback`, or maybe we should union `historyApiFallback` with `static` features?

Reply:
I think it makes sense to add it to `static` since it deals with serving static HTML content

Reply:
Yep, i think we should add the `historyApiFallback` option to static

Reply:
Good catch :+1: Thanks

Reply:
Although, is there any case where a user would want `static: false` with `historyApiFallback: true`? Or should we not worry about this edge case?

Reply:
I think you are right, you can use `historyApiFallback` without `static`

Reply:
I am afraid double `static` middleware reduce performance, but it is main for development, plus multiple `static` features in difference places look like hacky, can you describe the problem with `historyApiFallback` or maybe demonstrate?

Reply:
I'm looking into this now.

As a side note related to this, do you know why `middleware` is pushed to the features list 3 times? https://github.com/webpack/webpack-dev-server/blob/4ab1f21bc85cc1695255c739160ad00dc14375f1/lib/Server.js#L559

Reply:
Don't know, but I hope we will have tests on this, maybe we can look at git history why it was done

Reply:
@evilebottnawi I'm not really sure why it was added: https://github.com/webpack/webpack-dev-server/pull/1891/files

Reply:
@evilebottnawi After some searching I figured out that `contentBaseFiles` (https://github.com/webpack/webpack-dev-server/blob/4ab1f21bc85cc1695255c739160ad00dc14375f1/lib/Server.js#L562) feature was placed before and after `historyApiFallback` for this reason:

https://github.com/bripkens/connect-history-api-fallback/blob/master/examples/static-files-and-index-rewrite/README.md#configuring-the-middleware

Is it possible that the `middleware` feature is added multiple times for a similar reason? Changing it so that there is only 1 `middleware` pushed does not seem to break any tests, but maybe tests were not made for it?

Reply:
I think I will not mess with the `middleware` runnable feature, but will investigate it later and see if I can add some tests for having it get pushed multiple times.

I have made the 3 features: `static`, `serveIndex`, and `watchStatic`. It would be ideal to have only the `static` feature, but it is complicated by the `historyApiFallback` feature.

Reply:
It is very weird, because if we can't find something in `static` and `history` we should return `404`, not trying to do it again :confused: maybe I something don't understand

Reply:
@evilebottnawi It isn't a matter of finding something or returning a 404, but rather the priority of `static` vs `history` when they are both trying to serve something from the same path. I still don't know exactly why it works, but placing `static` middleware before and after seems to ensure that static files take priority over the history api fallback file.

Reply:
Yep, I see, very weird, but we can't do something here :disappointed: